core/src/main/scala/kafka/log/TimeIndex.scala

```diff
@@ -0,0 +1,198 @@
+        _entries += 1
+        _lastTimestamp = timestamp
+        _lastOffset = offset
+        require(_entries * entrySize == mmap.position, _entries + " entries but file position in index is " + mmap.position + ".")
+      }
+    }
+  }
+
+  /**
+   * Find the time index entries whose timestamp is less than the given timestamp and return a number of closest
+   * entries.
+   *
+   * @param targetTimestamp The timestamp to look up.
+   * @param numOffsets The maximum number of time index entries to return.
+   *
+   * @return A sequence of time index entries.
+   */
+  def lookup(targetTimestamp: Long, numOffsets: Int = 1): Seq[TimestampOffset] = {
+    maybeLock(lock) {
+      _entries match {
+        case 0 => Seq.empty
+        case _ =>
+          val idx = mmap.duplicate
+          val foundSlot = indexSlotFor(idx, targetKey = Some(targetTimestamp))
+          // The binary search will return the exact match when there is one. If the timestamp in the index entry is
+          // the same as target timestamp, we need to use the earlier entry.
+          val slot = {
+            if (foundSlot >= 0 && timestamp(idx, foundSlot) == targetTimestamp)
+              foundSlot - 1
+            else
+              foundSlot
+          }
+          if (slot == -1)
+            Seq.empty
+          else {
+            val entries = new ListBuffer[TimestampOffset]
+            for (i <- slot to math.max(0, slot - numOffsets + 1) by -1)
+              entries += parseEntry(idx, i).asInstanceOf[TimestampOffset]
+            entries.toSeq
+          }
+      }
+    }
+  }
+
+  override def truncate() = truncateToEntries(0)
+
+  /**
+   * Remove all entries from the index which have an offset greater than or equal to the given offset.
+   * Truncating to an offset larger than the largest in the index has no effect.
+   */
+  override def truncateTo(offset: Long) {
+    inLock(lock) {
+      val idx = mmap.duplicate
+      val slot = indexSlotFor(idx, targetValue = Some(offset))
+      /* There are 2 cases for choosing the new size
+       * 1) if there is no entry in the index <= the offset, delete everything
+       * 2) Otherwise delete everything larger than the closest smallest
+       */
+      val newEntries = if(slot < 0) 0 else slot + 1
+      truncateToEntries(newEntries)
+    }
+  }
+
+  /**
+   * Truncates index to a known number of entries.
+   */
+  private def truncateToEntries(entries: Int) {
+    inLock(lock) {
+      _entries = entries
+      mmap.position(_entries * entrySize)
+      _lastOffset = lastEntry.offset
+      _lastTimestamp = lastEntry.timestamp
+    }
+  }
+
+  override def sanityCheck() {
+    require(_entries == 0 || (_lastTimestamp >= timestamp(mmap, 0) && _lastOffset > baseOffset),
+      ("Corrupt time index found, index file (%s) has non-zero size but the last timestamp is %d and the " +
+        "first timestamp is %d")
+          .format(file.getAbsolutePath, _lastTimestamp, timestamp(mmap, 0)))
```

Comment:
The error message is only for timestamp, but the failure could be due to offset.